pkg/common/mesh_config.go

```diff
@@ -0,0 +1,159 @@
+package common
+
+import (
+	"fmt"
+
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/structpb"
+	istiomeshv1alpha1 "istio.io/api/mesh/v1alpha1"
+	istioapiv1alpha1 "istio.io/api/operator/v1alpha1"
+	iopv1alpha1 "istio.io/istio/operator/pkg/apis/istio/v1alpha1"
+	"istio.io/istio/pkg/util/protomarshal"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ExtAuthorizerName = "kuadrant-authorization"
+)
+
+// MeshConfigFromStruct Builds the Istio/OSSM MeshConfig from a compatible structure:
+//
+//	meshConfig:
+//	  extensionProviders:
+//	    - envoyExtAuthzGrpc:
+//	        port: <port>
+//	        service: <authorino-service>
+//	      name: kuadrant-authorization
+func MeshConfigFromStruct(structure *structpb.Struct) (*istiomeshv1alpha1.MeshConfig, error) {
+	if structure == nil {
+		return &istiomeshv1alpha1.MeshConfig{}, nil
+	}
+
+	meshConfigJSON, err := structure.MarshalJSON()
+	if err != nil {
+		return nil, err
+	}
+	meshConfig := &istiomeshv1alpha1.MeshConfig{}
+	// istiomeshv1alpha1.MeshConfig doesn't implement JSON/Yaml marshalling, only protobuf
+	if err = protojson.Unmarshal(meshConfigJSON, meshConfig); err != nil {
+		return nil, err
+	}
+
+	return meshConfig, nil
+}
+
+// MeshConfigFromString returns the Istio MeshConfig from a ConfigMap
+func MeshConfigFromString(config string) (*istiomeshv1alpha1.MeshConfig, error) {
+	meshConfig := &istiomeshv1alpha1.MeshConfig{}
+	err := protomarshal.ApplyYAML(config, meshConfig)
+	if err != nil {
+		return nil, err
+	}
+	return meshConfig, nil
+}
+
+// meshConfigToString returns the Istio MeshConfig as a string
+func meshConfigToString(config *istiomeshv1alpha1.MeshConfig) (string, error) {
+	configString, err := protomarshal.ToYAML(config)
+	if err != nil {
+		return "", err
+	}
+	return configString, nil
+}
+
+func MeshConfigToStruct(config *istiomeshv1alpha1.MeshConfig) (*structpb.Struct, error) {
+	configJSON, err := protojson.Marshal(config)
+	if err != nil {
+		return nil, err
+	}
+	configStruct := &structpb.Struct{}
+
+	if err = configStruct.UnmarshalJSON(configJSON); err != nil {
+		return nil, err
+	}
+	return configStruct, nil
+}
+
+func ExtensionProvidersFromMeshConfig(config *istiomeshv1alpha1.MeshConfig) (extensionProviders []*istiomeshv1alpha1.MeshConfig_ExtensionProvider) {
+	extensionProviders = config.ExtensionProviders
+	if len(extensionProviders) == 0 {
+		extensionProviders = make([]*istiomeshv1alpha1.MeshConfig_ExtensionProvider, 0)
+	}
+	return
+}
+
+func UpdateMeshConfig(configObject client.Object, updateFunc func(meshConfig *istiomeshv1alpha1.MeshConfig) bool) (bool, error) {
+	meshConfig, err := getMeshConfig(configObject)
+	if err != nil {
+		return false, err
+	}
+	if updateFunc(meshConfig) {
+		switch config := configObject.(type) {
```

Comment:
What do you think about having wrapper structs for the IstioOperator, for the ConfigMap and for ServiceMeshControlPlane resources that implement an interface to abstract these type-specific functions (`getMeshConfig`, `UpdateMesgConfig`) that otherwise require the type checking in a `switch` clause?